File: python/modular/Room.py
#
# Explore
#    - The Adventure Interpreter
#
# Copyright (C) 2006  Joe Peterson
#

import string

import World
import ExpUtil
from ItemContainer import *

class Room(ItemContainer):
    def __init__(self):
        ItemContainer.__init__(self)
        
        self.name = None
        self.desc = None
        self.desc_alt = None
        self.desc_ctrl = None
        self.neighbors = [None, None, None, None, None, None]
        self.original_neighbors = [None, None, None, None, None, None]
    
    def neighbor(self, direction):
        try:
            return self.neighbors[World.DIRECTIONS.index(direction)]
        except ValueError:
            return None
        
    def init_neighbor(self, direction, room):
        try:
            i = World.DIRECTIONS.index(direction)
        except ValueError:
            return
        
        self.neighbors[i] = room
        self.original_neighbors[i] = room
        
    def set_neighbor(self, direction, room):
        try:
            i = World.DIRECTIONS.index(direction)
        except ValueError:
            return

        self.neighbors[i] = room
        
    def revert_neighbor(self, direction):
        try:
            i = World.DIRECTIONS.index(direction)
        except ValueError:
            return

        self.neighbors[i] = self.original_neighbors[i]
        
    def block_way(self, direction):
        self.set_neighbor(direction, None)
        
    def make_way(self, direction, new_room):
        self.set_neighbor(direction, new_room)
        
    def neighbor_save_string(self, direction):
        try:
            i = World.DIRECTIONS.index(direction)
        except ValueError:
            return ""
        
        if self.neighbors[i] != self.original_neighbors[i]:
            if self.neighbors[i] != None:
                return self.neighbors[i]
            else:
                return "^"
        else:
            return ""
        
    def description(self):
        ctrl = "RC"
        out_desc = []
        
        if self.desc_ctrl != None:
            pos = self.desc_ctrl.find(",")
            if pos == -1:
                ctrl = self.desc_ctrl
            else:
                if self.desc_ctrl[-1] == "+":
                    ctrl = self.desc_ctrl[pos + 1:]
                else:
                    ctrl = self.desc_ctrl[:pos]

        if ctrl.find("R") != -1 and self.desc != None:
            out_desc.append(self.desc)

        if ctrl.find("C") != -1 and self.desc_alt != None:
            out_desc.append(self.desc_alt)

        if len(self.items) > 0:
            for item in self.items:
                out_desc.append("There is " + ExpUtil.a_or_an(item) + " " + item + " here.")

        return string.join(out_desc, '\n')

File: java/src/com/wildlava/explore/Room.java
//
// Explore
//    - The Adventure Interpreter
//
// Copyright (C) 2006  Joe Peterson
//

package com.wildlava.explore;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

class Room extends ItemContainer
{
   String name;
   String desc;
   String desc_alt;
   String desc_ctrl;
   List<String> fixed_objects;
   Map<String, String> neighbors;
   Map<String, String> original_neighbors;

   Room(World world)
   {
      super(world);

      neighbors = new HashMap<>();
      original_neighbors = new HashMap<>();
   }

   String description()
   {
      List<String> desc_strings = new ArrayList<>();
      String ctrl = "RC";

      if (desc_ctrl != null)
      {
         int pos = desc_ctrl.indexOf(",");

         if (pos == -1)
         {
            ctrl = desc_ctrl;
         }
         else
         {
            if (desc_ctrl.endsWith("+"))
            {
               ctrl = desc_ctrl.substring(pos + 1);
            }
            else
            {
               ctrl = desc_ctrl.substring(0, pos);
            }
         }
      }

      boolean r = (ctrl.indexOf("R") != -1 && desc != null);
      boolean c = (ctrl.indexOf("C") != -1 && desc_alt != null);

      if (r && c)
      {
         desc_strings.add(desc + "\\" + desc_alt);
      }
      else if (r)
      {
         desc_strings.add(desc);
      }
      else if (c)
      {
         desc_strings.add(desc_alt);
      }

      for (int i=0; i<items.size(); ++i)
      {
         String item = items.get(i);
         String item_lower = item.toLowerCase();

         if (World.trs_compat)
         {
            desc_strings.add("There is " + ExpUtil.aOrAn(item_lower) +
                             " " + item + " here.");
         }
         else
         {
            if (world.item_descs.containsKey(item))
            {
               desc_strings.add(world.item_descs.get(item));
            }
            else if (world.plural_items.contains(item))
            {
               desc_strings.add("There are some " + item_lower + " here.");
            }
            else if (world.mass_items.contains(item))
            {
               desc_strings.add("There is some " + item_lower + " here.");
            }
            else
            {
               desc_strings.add("There is " + ExpUtil.aOrAn(item_lower) +
                                " " + item_lower + " here.");
            }
         }
      }

      return ExpUtil.join(desc_strings, "\n");
   }

   boolean hasFixedObject(String item)
   {
      if (fixed_objects != null)
      {
         return fixed_objects.contains(item);
      }
      else
      {
         return false;
      }
   }

   String neighbor(String direction)
   {
      return neighbors.get(direction);
   }

   void initNeighbor(String direction, String room)
   {
      if (room == null)
      {
         neighbors.remove(direction);
         original_neighbors.remove(direction);
      }
      else
      {
         neighbors.put(direction, room);
         original_neighbors.put(direction, room);
      }
   }

   void setNeighbor(String direction, String room)
   {
      if (room == null)
      {
         neighbors.remove(direction);
      }
      else
      {
         neighbors.put(direction, room);
      }
   }

   void revertNeighbor(String direction)
   {
      String original_neighbor = original_neighbors.get(direction);
      if (original_neighbor == null)
      {
         neighbors.remove(direction);
      }
      else
      {
         neighbors.put(direction, original_neighbor);
      }
   }

   void blockWay(String direction)
   {
      setNeighbor(direction, null);
   }

   void makeWay(String direction, String new_room)
   {
      setNeighbor(direction, new_room);
   }

   String neighborSaveString(String direction)
   {
      String neighbor = neighbors.get(direction);
      String original_neighbor = original_neighbors.get(direction);

      if (neighbor == null && original_neighbor == null)
      {
         return "";
      }

      if (neighbor != null && neighbor.equals(original_neighbor))
      {
         return "";
      }

      if (neighbor == null)
      {
         return "^";
      }
      else
      {
         return neighbor;
      }
   }
}
